R/join_keys.R

```diff
@@ -0,0 +1,697 @@
+#'
+#' merge_join_keys(jk1, jk2)
+#' merge_join_keys(jk1, list(jk2))
+merge_join_keys.JoinKeys <- function(join_keys_obj, new_join_keys) {
+  assert_join_keys(join_keys_obj)
+
+  if (test_join_keys(new_join_keys)) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  lapply(new_join_keys, assert_join_keys_alike)
+
+  checkmate::assert_list(new_join_keys, types = c("JoinKeys"), min.len = 1)
+
+  for (el in new_join_keys) {
+    join_keys_obj <- modifyList(join_keys_obj, el)
+  }
+
+  logger::log_trace("JoinKeys keys merged.")
+  return(join_keys_obj)
+}
+
+#' Prints `JoinKeys`.
+#'
+#' @inheritParams base::print
+#' @return the `x` parameter
+#'
+#' @export
+print.JoinKeys <- function(x, ...) {
+  check_ellipsis(...)
+  keys_list <- x
+  my_parents <- parents(keys_list)
+  class(keys_list) <- "list"
+  if (length(keys_list) > 0) {
+    cat(sprintf(
+      "A JoinKeys object containing foreign keys between %s datasets:\n",
+      length(keys_list)
+    ))
+    # Hide parents
+    attr(keys_list, "__parents__") <- NULL
+    print.default(keys_list[sort(names(keys_list))])
+  } else {
+    cat("An empty JoinKeys object.")
+  }
+  invisible(x)
+}
+
+# -----------------------------------------------------------------------------
+#
+#
+# Helpers (non-exported)
+#
+
+#' Internal constructor
+#'
+#' @return an empty `JoinKeys` list
+#'
+#' @keywords internal
+new_join_keys <- function() {
+  structure(
+    list(),
+    class = c("JoinKeys", "list")
+  )
+}
+
+#' Helper function to add a new pair to a `JoinKeys` object
+#'
+#' @param join_keys_obj (`JoinKeys`) Object with existing pairs.
+#' @param join_key_obj (`JoinKeySet`) relationship pair to add.
+#'
+#' @keywords internal
+join_pair <- function(join_keys_obj, join_key_obj) {
+  assert_join_keys(join_keys_obj)
+  checkmate::assert_class(join_key_obj, "JoinKeySet")
+
+  dataset_1 <- dataset_1.JoinKeySet(join_key_obj)
+  dataset_2 <- dataset_2.JoinKeySet(join_key_obj)
+  keys <- keys.JoinKeySet(join_key_obj)
+
+  join_keys_obj[[dataset_1]][[dataset_2]] <- keys
+  join_keys_obj
+}
+
+#' Assert the JoinKeys class membership of an argument
+#' @inheritParams checkmate::assert_class
+#'
+#' @return `x` invisibly
+#'
+#' @keywords internal
+assert_join_keys <- function(x, .var.name = checkmate::vname(x), add = NULL) {
+  if (missing(x)) {
+    stop(sprintf("argument \"%s\" is missing, with no default", .var.name))
+  }
+
+  res <- check_join_keys(x)
+  checkmate::makeAssertion(x, res, var.name = .var.name, add)
+}
+
+#' @rdname assert_join_keys_alike
+#' @keywords internal
+check_join_keys <- function(x) {
+  checkmate::check_class(x, classes = c("JoinKeys", "list"))
+}
+
+#' @rdname assert_join_keys
+#' @keywords internal
+test_join_keys <- function(x) {
+  checkmate::makeTest(check_join_keys(x))
+}
+
+#' @rdname assert_join_keys
+#' @keywords internal
+expect_join_keys <- function(x, info = NULL, label = vname(x)) {
+  checkmate::makeExpectation(x, check_join_keys(x), info = info, label = label)
+}
```

Comment:
These sugars have very low value added. I'd remove them as `checkmate` calls are understandable enough (or even more ;)) 
```suggestion
```

Reply:
Makes sense, the idea was to reduce duplicated `c("JoinKeys", "list")` to also validate the content and symmetry, but that may be overkill.

Done in 497046bf (we can revert to a complex check later on if it makes sense)